cpp/ql/src/experimental/Security/CWE/CWE-193/InvalidPointerDeref.ql

```diff
@@ -0,0 +1,316 @@
+ * non-constant operand of the pointer-arithmetic operation.
+ */
+pragma[nomagic]
+predicate pointerAddInstructionHasBounds(
+  PointerAddInstruction pai, DataFlow::Node sink1, Instruction sink2, int delta
+) {
+  exists(Instruction right |
+    pointerAddInstructionHasOperands(pai, sink1.asInstruction(), right) and
+    bounded(right, sink2, delta)
+  )
+}
+
+/**
+ * Holds if `pai` is non-strictly upper bounded by `sink2 + delta` and `sink1` is the
+ * non-constant operand of the pointer-arithmetic operation.
+ */
+predicate isSinkImpl(
+  PointerAddInstruction pai, DataFlow::Node sink1, DataFlow::Node sink2, int delta
+) {
+  pointerAddInstructionHasBounds(pai, sink1, sink2.asInstruction(), delta)
+}
+
+/**
+ * Holds if `sink` is a sink for `InvalidPointerToDerefConf` and `i` is a `StoreInstruction` that
+ * writes to an address that non-strictly upper-bounds `sink`, or `i` is a `LoadInstruction` that
+ * reads from an address that non-strictly upper-bounds `sink`.
+ */
+predicate isInvalidPointerDerefSink(DataFlow::Node sink, Instruction i, string operation) {
+  exists(AddressOperand addr, int delta |
+    bounded(addr.getDef(), sink.asInstruction(), delta) and
+    delta >= 0 and
+    i.getAnOperand() = addr
+  |
+    i instanceof StoreInstruction and
+    operation = "write"
+    or
+    i instanceof LoadInstruction and
+    operation = "read"
+  )
+}
+
+/**
+ * A configuration to track flow from a pointer-arithmetic operation found
+ * by `AllocToInvalidPointerConf` to a dereference of the pointer.
+ */
+class InvalidPointerToDerefConf extends DataFlow3::Configuration {
+  InvalidPointerToDerefConf() { this = "InvalidPointerToDerefConf" }
+
+  override predicate isSource(DataFlow::Node source) { invalidPointerToDerefSource(_, source, _) }
+
+  override predicate isSink(DataFlow::Node sink) { isInvalidPointerDerefSink(sink, _, _) }
+}
+
+predicate invalidPointerToDerefSource(
+  PointerArithmeticInstruction pai, DataFlow::Node source, int delta
+) {
+  exists(ProductFlow::Configuration conf, DataFlow::PathNode p, DataFlow::Node sink1 |
+    p.getNode() = sink1 and
+    conf.hasFlowPath(_, _, p, _) and
+    isSinkImpl(pai, sink1, _, _) and
+    bounded(source.asInstruction(), pai, delta) and
+    delta >= 0
+  )
+}
+
+newtype TMergedPathNode =
+  // The path nodes computed by the first projection of `AllocToInvalidPointerConf`
+  TPathNode1(DataFlow::PathNode p) or
+  // The path nodes computed by `InvalidPointerToDerefConf`
+  TPathNode3(DataFlow3::PathNode p) or
+  // The read/write that uses the invalid pointer identified by `InvalidPointerToDerefConf`.
+  // This one is needed because the sink identified by `InvalidPointerToDerefConf` is the
+  // pointer, but we want to raise an alert at the dereference.
+  TPathNodeSink(Instruction i) {
+    exists(DataFlow::Node n |
+      any(InvalidPointerToDerefConf conf).hasFlow(_, n) and
+      isInvalidPointerDerefSink(n, i, _)
+    )
+  }
```

Comment:
So the overall structure is something like:
```
1. the allocation, e.g. p = malloc(size)
 v
2. 'end pointer' is calculated, e.g. q = p + size
(critically this is not just any pointer addition, its adding the allocation size to the allocation base)
 v
3. end pointer is accessed, e.g. *(q + 1)
```
I think this query would be a lot easier to read and understand if the above (or a corrected version of it) was established near the top, along with terminology (e.g. "allocation base", "allocation size" and "end pointer" I've used above) that can replace much of the anonymous stuff like `sink2` in the predicates that follow.

Reply:
I've added such a comment in 60b9a373ece9d1ec451c05422fc48f127795aa53. Let me know if you think it's more clear. I've kept the names `sink1` and `sink2` since I find it useful to know that a parameter will be used as the (first or second) sink of the product flow configuration. I've added some examples to the predicates that should hopefully explain the role of these somewhat anonymous parameters better.

Reply:
That's missing a key point that the step from 2 to 3 is a relative bound, not (necessarily) a direct copy. That means it can handle things like the following:
```
int *arr = malloc(size);
int *end = arr + size;
for(int *p = arr; p <= end; p++) {
  use(*p);
}
```

Reply:
Good point. I've gone with your example (and expanded the explanation) in dc00643ad100435ca6520397f4af3342da714074.